candlepin/candlepin
# DBunit - PoC

This POC tried add DbUnit utility to imporve isolation and remove verbosity in DB (Curator's) tests.

 Pros: 
+ test are more isolated (instead curators we use DbUnit)
       
+ test are cleaner and less verbose
        
 Cons:

- split logic into more files (Java and datasets) => harder to understand tests  
   
- creating curator's test are not easy to write in maintainable way (small sets of datasets files).
  Even when we don't care about maintability these tests requiere more time to write them,
  which violate testing rule: Unit test should be easy to write and run 

- I got CyclicTablesDependencyException when tried tranform PoolManagerFunctionalDbUnitTest
  to solve this problem you have to split dataset to two datasets + turn off referetial contraint check. 
    First will do the inserts, having problematic fields in null.
    The second dataset will then update the records, defining dependencies according to our needs.

- MAINTAIBILTY:

- devel needs to know DB schema and think about it when creating tests (more overhead)
- more tests could use same dataset file (maintability => minimize data sets), 
  so when change in file is required by new test, develop must check if don't break other tests, 
  which use the datset
    
- before test devel should check if dataset doesn't alread exist
  we should rather use existing datasets and compositing them together to keep count of data set file low,
  so devel need to check datasets before creating test (more time)
 
- PERFORMANCE:

  Test preparation:
  CP way:
    1. create transient obejct
    2. persist object and return reference to instance  // in-memoery DB hit
    3. ... testing with obejct 

  DbUnit way:
    1. load dataset file                                // READ file
    2. INSERT data by DBUnit tool                       // in-memory DB hit
    3. read object trough entityManager                 // in-memory DB hit
        (we have to read entity in case when entity will be used to build another entity,
        which will be stored as well -> we don't have cascade on Entities, so we have do it this way)

  - but 1. can be treated by (e. g. when dataset which can be reused by tests, I can read file only once
    for testsuit and use copy of dataset in tests)

===================================================  
      

=== Summary PoC-DbUnit ====
My recommendation : I would radher don't use DbUnit becase of more cons then pros.

Isolation: don't use curators methods, rather use directly EnityManager 
   *e. g. for importing data into db use JPA/Hibernate/SQL instead of DAO method which is tested

Verbosity: hide DB preparation code into private methods with appropriate names
===========================
